### PR TITLE
1.13 support, multiple configurations

### PR DESCRIPTION
**Pull request by FrankHeijden**
- 1.13 support -> Refactored ScanRegionTask into a cleaner looking class, and added the 1.13 blocks to not look for in the scan score. #45 
- Added Emerald ores to count in the resource score.
- Added option to disable animal-teleportation #14 
- All signs on the post are now configurable.
- Fix for hoppers under MinimumRegionPostY #43 
- Added option to thin Iron Golems too #28 
- The whole mid-regio-post thingy is configurable now. Setup of #29 is now possible:
screenshot 2018-07-24 at 23 55 47

### CODE REVIEW

**Comment by NatoBoram** [comment]
Are you able to upload a compiled `.jar` in the meantime? :D

**Comment by FrankHeijden** [comment]
Sure, please bare in mind it's only compatible with 1.13, not with 1.12.2 or lower.
https://frankheijden.stackstorage.com/s/tYQnib5yLFUfaZm

**Comment by RoboMWM** [comment]
Lol, both you and @zedwick submitting 1.13 PRs for both projects, lol. I'll look into these in a bit (just got done reviewing the GP ones lol), also would be interested in what @Developaws thinks of either.

My primary critique here though is bundling multiple features into one PR. Makes it harder to review the diff as well as only having a single thread to discuss all of these features instead of one per feature.

**Comment by FrankHeijden** [comment]
You can test the 1.13 jar I posted, everything works hehe

**Comment by FrankHeijden** [comment]
+ I bundled them, but in different pushes so that makes it a bit easier :x

**Comment by RoboMWM** [comment]
> but in different pushes so that makes it a bit easier

Unfortunately not the case, since the end diff (branch) is what matters in the end. Commits could include other minor changes to other things such as feedback and typos and etc., also wouldn't be able to see a change/fix you made in another commit when reviewing a single commit, etc.

**Comment by ghost** [comment]
Code looks good, and fixes a nice chunk of issues.
Code definately a lot cleaner than Zedwicks, so promoting this commit instead :)
Code does work as tested.

**Review by ghost — APPROVED** [review]

**Comment by FrankHeijden** [comment]
Compiled jar: https://frankheijden.stackstorage.com/s/ws1L9FWT7TKcUxV

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java [hunk 1]
```diff
@@ -52,21 +52,13 @@ public void onEntityToggleFlight(EntityToggleGlideEvent event)
             event.setCancelled(true);
         }
     }
-   
-    @EventHandler(ignoreCancelled = true)
-    public void onPlayerMove(PlayerMoveEvent event) {
```
Comment (RoboMWM) [inline src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java:57]: Any reason why you removed this event handler? @Developaws added this in https://github.com/TechFortress/PopulationDensity/pull/40 (also this is a good reason why one should title PRs properly, took me a while to find the PR lol)

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java [hunk 2]
```diff
@@ -52,21 +52,13 @@ public void onEntityToggleFlight(EntityToggleGlideEvent event)
...
-    }
 
-    @EventHandler(ignoreCancelled = true)
+    @EventHandler(priority = EventPriority.HIGHEST)
     public void onEntityDamage(EntityDamageEvent event)
     {
         Entity entity = event.getEntity();
         //when an entity has fall damage immunity, it lasts for only ONE fall damage check
-        if(event.getCause() == EntityDamageEvent.DamageCause.FALL)
+        if(event.getCause().equals(EntityDamageEvent.DamageCause.FALL))
```
Comment (RoboMWM) [inline src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java]: Standard for comparing enums is to use `==`

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java [hunk 3]
```diff
@@ -91,21 +83,17 @@ void makeEntityFallDamageImmune(LivingEntity entity)
         {
             Player player = (Player) entity;
             if(player.getGameMode() == GameMode.CREATIVE || player.getGameMode() == GameMode.SPECTATOR) return;
-            player.setGliding(false);
```
Comment (RoboMWM) [inline src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java]: Hmm, this is no longer a thing? Have elytras changed drastically in 1.13 that there is no such gliding ability?

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java [hunk 4]
```diff
@@ -18,11 +18,7 @@
 
 package me.ryanhamshire.PopulationDensity;
 
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Random;
+import java.util.*;
```
Comment (RoboMWM) [inline src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:21]: Don't use wildcard imports, fix your IDE to not do this.

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java [hunk 5]
```diff
@@ -126,14 +139,14 @@ public void onItemDespawn (ItemDespawnEvent event)
 		ItemStack item = ((Item)entity).getItemStack();
 		
 		//only care about saplings
-		if(item.getType() != Material.SAPLING) return;
+		if(saplings.contains(item.getType())) return;
```
Comment (RoboMWM) [inline src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:134]: As aforementioned, standard for comparing enums is `==`

File: src/me/ryanhamshire/PopulationDensity/PopulationDensity.java [hunk 6]
```diff
@@ -26,15 +26,7 @@
...
-import org.bukkit.ChatColor;
-import org.bukkit.Chunk;
-import org.bukkit.ChunkSnapshot;
-import org.bukkit.GameMode;
-import org.bukkit.Location;
-import org.bukkit.Material;
-import org.bukkit.OfflinePlayer;
-import org.bukkit.Sound;
-import org.bukkit.World;
+import org.bukkit.*;
```
Comment (RoboMWM) [inline src/me/ryanhamshire/PopulationDensity/PopulationDensity.java]: Also as aforementioned, no wildcard imports please c:

**Review by RoboMWM — DISMISSED** [review]
Just a couple changes to revert, and also a question on the gliding/fall damage thing.

**Comment by RoboMWM** [comment]
Oh, and do you have a converter for the region post values? Or at least something to catch the old config values (which contain ints for the now-legacy Material IDs)?

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java (hunk 1 above)
Comment (FrankHeijden) [reply src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java:57]: this was the cause of the damage still being present when teleporting to a region

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java (hunk 2 above)
Comment (FrankHeijden) [reply src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java]: doesn't matter which one you use

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java (hunk 4 above)
Comment (FrankHeijden) [reply src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:21]: what's the problem with that lol

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java (hunk 3 above)
Comment (FrankHeijden) [reply src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java]: will revert this back, seemed unneccessary to me but I see now why this was there lol.

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java (hunk 2 above)
Comment (RoboMWM) [reply src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java]: That's true, but would just like to keep consistency is all. Also makes it somewhat visually easier to identify an enum comparison IMO

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java (hunk 5 above)
Comment (FrankHeijden) [reply src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:134]: is not an enum

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java (hunk 1 above)
Comment (RoboMWM) [reply src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java:57]: Hmm, so playermoveevent would fire when player is on ground but before receiving damage? Makes sense then; PR #40 then is flawed :p

**Review by FrankHeijden — COMMENTED** [review]
not really, because that would be +3hrs of work to add. I think 1.13 is a great way of removing the ID's completely. + the user would need to convert all this by themself because of the new configurations added to the regionpost design - so it wouldn't matter. Also, defaults are assigned so the regionposts will work with just a drag and drop of the jar file.

File: src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java (hunk 2 above)
Comment (FrankHeijden) [reply src/me/ryanhamshire/PopulationDensity/DropShipTeleporter.java]: ¯\_(ツ)_/¯ will revert back xd

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java (hunk 4 above)
Comment (RoboMWM) [reply src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:21]: Easier to identify what the class is using, helpful when I'm debugging at times, especially when looking at repos I'm unfamiliar with (i.e. helpful for future maintainers/contributors). https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java (hunk 5 above)
Comment (RoboMWM) [reply src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:134]: Whoops, read it too fast lol.

**Comment by RoboMWM** [comment]
> not really, because that would be +3hrs of work to add.

Then at least catch an unconverted config and notify in server if they still have old IDs?

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java (hunk 4 above)
Comment (FrankHeijden) [reply src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:21]: but intellij got you covered with that:
```
org.bukkit.block.data.type.Sign wall = (org.bukkit.block.data.type.Sign) block.getBlockData();
wall.setRotation(blockFace);
block.setBlockData(wall);

Sign s = (Sign)block.getState();
```
Same variable Sign, but assigned differently

**Comment by FrankHeijden** [comment]
> Then at least catch an unconverted config and notify in server if they still have old IDs?

Sure, but I think it's better when we post this update to just type in bold above the update **PLEASE UPDATE YOUR CONFIG** xD

File: src/me/ryanhamshire/PopulationDensity/PopulationDensity.java [hunk 7]
```diff
@@ -1454,28 +1444,14 @@ static void removeMonstersAround(Location location)
...
-        }
-        catch(NumberFormatException e)
-        {
-            PopulationDensity.AddLogEntry("Couldn't understand config entry '" + string + "'.  Use format 'id:data'.");
-        }
-        
-        return null;
+        Material material = Material.getMaterial(string);
+        if (material == null) {
+			AddLogEntry("Error: Couldn't resolve material \""+string+"\". Please update your config.yml.");
```
Comment (FrankHeijden) [inline src/me/ryanhamshire/PopulationDensity/PopulationDensity.java]: It's already here btw

File: src/me/ryanhamshire/PopulationDensity/PopulationDensity.java [hunk 8]
```diff
@@ -1454,28 +1465,14 @@ static void removeMonstersAround(Location location)
...
```
Comment (FrankHeijden) [inline src/me/ryanhamshire/PopulationDensity/PopulationDensity.java:1483]: It's already here

**Comment by ghost** [comment]
As for the changes that @RoboMWM suggested, a lot of the formatting stuff isn't /really/ required, as it's only temporary till I do a major overhaul of this system.
But as per everything else, it's mostly good :P Silly @RoboMWM

**Comment by NatoBoram** [comment]
> Pagina niet gevonden
> 
> Het opgegeven document is niet gevonden op deze server.
> Klik hier om terug te keren

Perhaps you can upload it directly to GitHub?

**Comment by FrankHeijden** [comment]
[PopulationDensity.jar.zip](https://github.com/TechFortress/PopulationDensity/files/2229704/PopulationDensity.jar.zip)

File: src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java (hunk 4 above)
Comment (RoboMWM) [reply src/me/ryanhamshire/PopulationDensity/EntityEventHandler.java:21]: I'm not talking about specific usages/conflicts in the code - I just mean when I look at a class file, seeing what it imports tells me at a glance what it does/works with.

**Comment by RoboMWM** [comment]
> we post this update to just type in bold above the update

You'd be surprised how many don't read - and with the changes to BukkitDev, you're just given a download, no changelog page first.

File: src/me/ryanhamshire/PopulationDensity/PopulationDensity.java (hunk 8 above)
Comment (RoboMWM) [reply src/me/ryanhamshire/PopulationDensity/PopulationDensity.java:1483]: Was thinking along the lines of being a message after startup in log, but at least the more diligent administrators will see this.

**Comment by RoboMWM** [comment]
>  a lot of the formatting stuff isn't /really/ required

True, but most of the changes I commented on don't change anything else. As if it were some unnecessary change. And it's good to at least maintain _some_ level of consistency.

**Comment by RoboMWM** [comment]
Also, don't know how important emeralds are (afaik, they're solely for villager trading..?), and from what I recall they don't spawn in pockets like the other ores? Either way, will probably need some testing on it... will have to see the natural spawn rates for emerald ores.

**Comment by RoboMWM** [comment]
Was about to pull this and then realized you've been adding commits still, so I'll wait until you think it's ready to be pulled c: (Then if you wish to keep adding features you can do that in new PRs.)

**Comment by FrankHeijden** [comment]
This PR is finished, I thought emerald was a great addition as it’s somewhat harder to find than diamond. It only occurs in a mountains biome and the wiki says it can occur as high as gold can. So maybe ill change the resource vale to the gold value.

**Comment by RoboMWM** [comment]
> somewhat harder to find than diamond. It only occurs in a mountains biome

> can occur as high as gold can

If the ore is specific to a certain biome, it doesn't seem to make a lot of sense to include it in the resource count. Remember, the resources score is there to determine if a place has been mined out by existing players. Wood is included since that's a resource every player starting out needs (and also an easy way to prevent spawning players in an ocean). Emeralds are late-game resources, and as you said, are only found in a specific biome. Other than it being an ore like the other ores, I don't see any reason why this should be included. (Perhaps if they were present in every biome like diamonds, I'd see a rationale.)

**Comment by RoboMWM** [comment]
Since the resource score is cumulative this is okay, otherwise I'd argue that it's a late-game resource that can't be found anywhere except in one biome.

**Pull request merged**